File: multitype/src/main/java/com/senierr/adapter/internal/MultiTypeAdapter.java
package com.senierr.adapter.internal;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 多类型适配器
 *
 * @author zhouchunjie
 * @date 2017/9/25
 */
public class MultiTypeAdapter extends RecyclerView.Adapter<ViewHolder> {

    private @Nullable RecyclerView recyclerView;
    private @NonNull List<Object> data;
    // 视图集合：对应的Index即为ItemViewType
    private List<ViewHolderWrapper<?>> wrapperList;
    // 数据类型对应的连接器
    private Map<Class<?>, Linker<?>> linkerMap;

    public MultiTypeAdapter() {
        this(new ArrayList<>());
    }

    public MultiTypeAdapter(@NonNull List<Object> data) {
        this.data = data;
        wrapperList = new ArrayList<>();
        linkerMap = new HashMap<>();
    }

    @Override @SuppressWarnings("unchecked")
    public final int getItemViewType(int position) {
        Object item = data.get(position);
        // 获取数据类型对应的连接器
        Linker linker = linkerMap.get(item.getClass());
        if (linker == null) {
            throw new UnregisteredException(item.getClass());
        }
        // 根据连接器，获取对应的视图类型
        ViewHolderWrapper viewHolderWrapper = linker.getItemViewType(item);
        // 根据视图类型转换为Type
        int itemViewType = wrapperList.indexOf(viewHolderWrapper);
        if (itemViewType < 0 || itemViewType > wrapperList.size() - 1) {
            throw new UnregisteredException(viewHolderWrapper.getClass());
        }
        return itemViewType;
    }

    @Override
    public final ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ViewHolderWrapper<?> viewHolderWrapper = wrapperList.get(viewType);
        return viewHolderWrapper.generateViewHolder(parent, this);
    }

    @Override @SuppressWarnings("unchecked")
    public final void onBindViewHolder(ViewHolder holder, int position, @NonNull List<Object> payloads) {
        Object item = data.get(position);
        ViewHolderWrapper viewHolderWrapper = wrapperList.get(holder.getItemViewType());
        viewHolderWrapper.onBindViewHolder(holder, item, payloads);
    }

    @Override @Deprecated
    public final void onBindViewHolder(@NonNull ViewHolder holder, int position) {}

    @Override
    public final int getItemCount() {
        return data.size();
    }

    @Override @SuppressWarnings("unchecked")
    public final long getItemId(int position) {
        ViewHolderWrapper viewHolderWrapper = wrapperList.get(getItemViewType(position));
        return viewHolderWrapper.getItemId(data.get(position));
    }

    @Override
    public final void onViewRecycled(@NonNull ViewHolder holder) {
        super.onViewRecycled(holder);
        ViewHolderWrapper viewHolderWrapper = wrapperList.get(holder.getItemViewType());
        viewHolderWrapper.onViewRecycled(holder);
    }

    @Override
    public final boolean onFailedToRecycleView(ViewHolder holder) {
        ViewHolderWrapper viewHolderWrapper = wrapperList.get(holder.getItemViewType());
        return viewHolderWrapper.onFailedToRecycleView(holder);
    }

    @Override @SuppressWarnings("unchecked")
    public final void onViewAttachedToWindow(@NonNull ViewHolder holder) {
        super.onViewAttachedToWindow(holder);
        if (recyclerView == null) {
            return;
        }

        ViewHolderWrapper viewHolderWrapper = wrapperList.get(holder.getItemViewType());
        viewHolderWrapper.onViewAttachedToWindow(holder);

        ViewGroup.LayoutParams lp = holder.itemView.getLayoutParams();
        if (lp != null && lp instanceof StaggeredGridLayoutManager.LayoutParams) {
            StaggeredGridLayoutManager.LayoutParams p = (StaggeredGridLayoutManager.LayoutParams) lp;
            final RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
            if (layoutManager == null) return;
            StaggeredGridLayoutManager staggeredGridLayoutManager = (StaggeredGridLayoutManager) layoutManager;
            int position = holder.getLayoutPosition();
            int spanSize = viewHolderWrapper.getSpanSize(data.get(position));
            p.setFullSpan(spanSize >= staggeredGridLayoutManager.getSpanCount());
        }
    }

    @Override
    public final void onViewDetachedFromWindow(@NonNull ViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        ViewHolderWrapper viewHolderWrapper = wrapperList.get(holder.getItemViewType());
        viewHolderWrapper.onViewDetachedFromWindow(holder);
    }

    @Override @SuppressWarnings("unchecked")
    public final void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        this.recyclerView = recyclerView;

        final RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
        if (layoutManager instanceof GridLayoutManager) {
            final GridLayoutManager gridManager = ((GridLayoutManager) layoutManager);
            gridManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    ViewHolderWrapper viewHolderWrapper = wrapperList.get(getItemViewType(position));
                    int spanSize = viewHolderWrapper.getSpanSize(data.get(position));
                    int spanCount = gridManager.getSpanCount();
                    return spanSize > spanCount ? spanCount : spanSize;
                }
            });
        }

        for (ViewHolderWrapper<?> viewHolderWrapper : wrapperList) {
            viewHolderWrapper.onAttachedToRecyclerView(recyclerView);
        }
    }

    @Override
    public final void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        this.recyclerView = null;

        for (ViewHolderWrapper<?> viewHolderWrapper : wrapperList) {
            viewHolderWrapper.onDetachedFromRecyclerView(recyclerView);
        }
    }

    @NonNull
    public List<Object> getData() {
        return data;
    }

    public void setData(@NonNull List<Object> data) {
        this.data = data;
    }

    public final <T> void register(@NonNull Class<T> cls, @NonNull final ViewHolderWrapper<T> viewHolderWrapper) {
        new RegisterHelper<>(cls, viewHolderWrapper).with(new Linker<T>() {
            @NonNull
            @Override
            public ViewHolderWrapper<T> getItemViewType(@NonNull T item) {
                return viewHolderWrapper;
            }
        });
    }

    @NonNull @SafeVarargs @CheckResult
    public final <T> RegisterHelper<T> register(@NonNull Class<T> cls, @NonNull ViewHolderWrapper<T>... viewHolderWrappers) {
        return new RegisterHelper<>(cls, viewHolderWrappers);
    }

    public final class RegisterHelper<T> {

        private Class<T> cls;
        private ViewHolderWrapper<T>[] viewHolderWrappers;

        @SafeVarargs
        RegisterHelper(Class<T> cls, ViewHolderWrapper<T>... viewHolderWrappers) {
            this.cls = cls;
            this.viewHolderWrappers = viewHolderWrappers;
        }

        public final void with(@NonNull Linker<T> linker) {
            for (ViewHolderWrapper<T> viewHolderWrapper : viewHolderWrappers) {
                viewHolderWrapper.setMultiTypeAdapter(MultiTypeAdapter.this);
                wrapperList.add(viewHolderWrapper);
            }
            linkerMap.put(cls, linker);
        }
    }
}

File: multitype/src/main/java/com/senierr/adapter/internal/MultiTypeAdapter.kt
package com.senierr.adapter.internal

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import android.view.ViewGroup

/**
 * 多类型适配器
 *
 * @author zhouchunjie
 * @date 2017/9/25
 */
class MultiTypeAdapter(var data: MutableList<Any> = mutableListOf()) : RecyclerView.Adapter<ViewHolder>() {

    private var recyclerView: RecyclerView? = null
    private val itemBinders = mutableListOf<ItemBinder<*>>()
    private val wrapperCache = mutableListOf<ViewHolderWrapper<*>>()
    // 自定义类型匹配
    private var viewTypeLinker: ViewTypeLinker? = null

    override fun getItemViewType(position: Int): Int {
        val item = data[position]

        // 优先自定义视图类型
        val linker = viewTypeLinker
        if (linker != null) {
            return linker.getItemViewType(item)
        }

        val index = indexOf(item)
        if (index == -1) throw UnregisteredException(item.javaClass)
        return index
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val viewHolderWrapper = getViewHolderWrapper<Any>(viewType)
        return viewHolderWrapper.onCreateViewHolderInternal(parent)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
        val item = data[position]
        val viewHolderWrapper = getViewHolderWrapper<Any>(holder.itemViewType)
        viewHolderWrapper.onBindViewHolder(holder, item, payloads)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    }

    override fun getItemCount(): Int = data.size

    override fun onViewAttachedToWindow(holder: ViewHolder) {
        super.onViewAttachedToWindow(holder)
        val viewHolderWrapper = getViewHolderWrapper<Any>(holder.itemViewType)
        viewHolderWrapper.onViewAttachedToWindow(holder)

        recyclerView?.let {
            val lp = holder.itemView.layoutParams
            val layoutManager = it.layoutManager
            if (lp != null && lp is StaggeredGridLayoutManager.LayoutParams
                    && layoutManager != null && layoutManager is StaggeredGridLayoutManager) {
                val position = holder.layoutPosition
                val spanSize = viewHolderWrapper.getSpanSize(data[position])
                lp.isFullSpan = spanSize >= layoutManager.spanCount
            }
        }
    }

    override fun onViewDetachedFromWindow(holder: ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        val viewHolderWrapper = getViewHolderWrapper<Any>(holder.itemViewType)
        viewHolderWrapper.onViewDetachedFromWindow(holder)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView

        val layoutManager = recyclerView.layoutManager
        if (layoutManager != null && layoutManager is GridLayoutManager) {
            layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    val viewHolderWrapper = getViewHolderWrapper<Any>(getItemViewType(position))
                    val spanSize = viewHolderWrapper.getSpanSize(data[position])
                    val spanCount = layoutManager.spanCount
                    return if (spanSize > spanCount) spanCount else spanSize
                }
            }
        }

        for (viewHolderWrapper in wrapperCache) {
            viewHolderWrapper.onAttachedToRecyclerView(recyclerView)
        }
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        this.recyclerView = null

        for (viewHolderWrapper in wrapperCache) {
            viewHolderWrapper.onDetachedFromRecyclerView(recyclerView)
        }
    }

    /**
     * 获取数据项对应的封装器的索引
     */
    @Suppress("UNCHECKED_CAST")
    private fun indexOf(item: Any): Int {
        var index = 0
        itemBinders.forEach {
            if (it.type == item::class.java) {
                val delegation = it.delegation as Delegation<Any>
                val wrapperType = delegation.getWrapperType(item)
                for (i in 0 until it.viewHolderWrappers.size) {
                    if (it.viewHolderWrappers[i]::class.java == wrapperType) {
                        index += i
                        return index
                    }
                }
            } else {
                index += it.viewHolderWrappers.size
            }
        }
        return -1
    }

    /**
     * 获取封装器
     */
    @Suppress("UNCHECKED_CAST")
    private fun <T> getViewHolderWrapper(itemViewType: Int): ViewHolderWrapper<T> {
        // 优先自定义视图类型
        val linker = viewTypeLinker
        if (linker != null) {
            return linker.getViewHolderWrapper(itemViewType) as ViewHolderWrapper<T>
        }
        return wrapperCache[itemViewType] as ViewHolderWrapper<T>
    }

    /**
     * 重置封装器
     */
    private fun resetWrapperCache() {
        wrapperCache.clear()
        itemBinders.forEach {
            wrapperCache.addAll(it.viewHolderWrappers)
        }
    }

    /**
     * 解除注册
     */
    fun <T> unregister(type: Class<T>) {
        val iterator = itemBinders.iterator()
        while (iterator.hasNext()) {
            val item = iterator.next()
            if (item.type == type) {
                iterator.remove()
            }
        }
        resetWrapperCache()
    }

    /**
     * 注册
     */
    fun <T> register(type: Class<T>, vararg viewHolderWrapper: ViewHolderWrapper<T>, delegation: Delegation<T>) {
        // 解除相同数据类型的注册
        unregister(type)
        // 注册
        itemBinders.add(ItemBinder(type, mutableListOf(*viewHolderWrapper), delegation))
        viewHolderWrapper.forEach {
            it.onRegister(this)
        }
        // 重置缓存
        resetWrapperCache()
    }

    /**
     * 注册
     */
    fun <T> register(type: Class<T>, viewHolderWrapper: ViewHolderWrapper<T>) {
        register(type, viewHolderWrapper, delegation = object : Delegation<T> {
            override fun getWrapperType(item: T): Class<out ViewHolderWrapper<T>> {
                return viewHolderWrapper::class.java
            }
        })
    }

    /**
     * 注册自定义匹配规则
     */
    fun register(viewHolderWrapper: List<ViewHolderWrapper<*>>, viewTypeLinker: ViewTypeLinker) {
        wrapperCache.clear()
        wrapperCache.addAll(viewHolderWrapper)
        wrapperCache.forEach {
            it.onRegister(this)
        }
        this.viewTypeLinker = viewTypeLinker
    }

    inline fun <reified T> register(viewHolderWrapper: ViewHolderWrapper<T>) {
        register(T::class.java, viewHolderWrapper)
    }

    inline fun <reified T> register(vararg viewHolderWrapper: ViewHolderWrapper<T>, delegation: Delegation<T>) {
        register(T::class.java, *viewHolderWrapper, delegation = delegation)
    }

    inline fun <reified T> register(vararg viewHolderWrapper: ViewHolderWrapper<T>,
                                    crossinline delegation: (item: T) -> Class<out ViewHolderWrapper<T>>) {
        register(T::class.java, *viewHolderWrapper, delegation = object : Delegation<T> {
            override fun getWrapperType(item: T): Class<out ViewHolderWrapper<T>> {
                return delegation.invoke(item)
            }
        })
    }
}